File: leetcode/剑指offer/剑指 Offer 09. 用两个栈实现队列/剑指offer09.cpp
#include<iostream>
#include <stack>
using namespace std;

//CQueue创建双栈列
class CQueue {
    stack<int> stack1,stack2;
public:
    //  声明空的栈列
    CQueue() {
        while (!stack1.empty()) {
            stack1.pop();
        }
        while (!stack2.empty()) {
            stack2.pop();
        }
    }
    //push添加栈元素
    //先入stack1中，先入后出
    void appendTail(int value) {
        stack1.push(value);
    }
    // 执行队列删除操作
    //将stack1中的先入后出顺序转移到stack2中的先进先出顺序
    int deleteHead() {
        // 如果第二个栈为空
        if (stack2.empty()) {
            while (!stack1.empty()) {           //while循环，将stack1中所有元素转变到stack2中
                stack2.push(stack1.top());
                stack1.pop();
            }
        } 
        //执行删除操作
        if (stack2.empty()) {
            return -1;
        } else {
            int deleteItem = stack2.top();  // = stack1最底部元素
            stack2.pop();
            return deleteItem;  //返回删除值
        }
    }
};

File: leetcode/题库整理/剑指offer/剑指 Offer 09. 用两个栈实现队列/剑指offer09.py
class CQueue:
    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    #定义添加元素操作
    def appendTail(self, value: int) -> None:
        # # 当输入为stack1时，获取此时stack2的值
        # while self.stack1:
        #     self.stack2.append(self.stack1.pop())

        # 将value加入到stack1
        self.stack1.append(value)
        
        # # 当输入为stack2时，需要转移到stack1中存储
        # while self.stack2:
        #     self.stack1.append(self.stack2.pop())
        # return self.stack1

    #定义删除元素操作
    def deleteHead(self) -> int:    #返回int
        #当目标为stack2，且不为空，返回stack2头元素
        if self.stack2: return self.stack2.pop()

        if not self.stack1: return -1   #当目标为stack1且为空，返回-1

        #当目标为stack1且不为空，要想达到先进先出，需要使用两个栈转移
        while self.stack1:
            self.stack2.append(self.stack1.pop())   #stack1 -> stack2
        return self.stack2.pop()    #返回stack2头元素   